### DISCUSSION TITLE
No encryption key provided

### DISCUSSION THREAD

**Discussion by m1001101** [teslamate-org/teslamate#2798] [Q&A] [answered]
Hello, I've added the key in the yml file and Pi environment via /etc/profile, rebooted, but I always get the message "No encryption key provided".

![image](https://user-images.githubusercontent.com/7558556/180206455-1b5c33f9-68b2-4629-9526-6b1a3226bd8f.png)

from the message I don't understand where the key goes.

Thanks!

M

**Comment by ktsenvolt** [comment, marked as answer]
You need to do the following commands:
docker-compose pull
docker-compose up -d

**Reply by m1001101** [reply]
ok after insert ENCRYPTION_KEY in raspbian environment (/etc/profile), and docker-compose pull && docker-compose up -d it works!

Thanks!

**Comment by LelandSindt** [comment]
@m1001101 Have a look at the release notes for 1.27.0: https://github.com/adriankumpf/teslamate/releases/tag/v1.27.0

Assuming you are running in docker using docker-compose you will need to edit your `docker-compose.yaml` file to include the `ENCRYPTION_KEY` environment variable.

**Reply by m1001101** [reply]
Yes, I have read the notes but I don't know anything about docker-compose and I didn't know had to use the two commands (docker-compose pull; docker-compose up -d) to apply the change

**Comment by Etigro** [comment]
<img width="973" alt="Capture d’écran 2022-08-02 à 11 57 46" src="https://user-images.githubusercontent.com/90093992/182347743-07fdcb47-ffe2-4fd5-bfbe-585e4708690d.png">

Eventhough I did "pull" and "compose up" I still get the connection error. I did edit the docker-compose.yml file and the docker-compose config command shows the ENCRYPTION_KEY variable. Any idea ?

**Reply by Etigro** [reply]
I figured I need to generate the Tesla token again as the initial connection failed. Given the correct credential enabled a successful connection. All good now.

**Reply by ventz** [reply]
Just wanted to add that @Etigro's answer is actually the problem for those upgrading.

If you go from "pre 1.26" -> "1.27", even if the encryption key is set in the docker-compose file, and you pull + restart, you still get the message.

Adding the new API access+refresh tokens and *then* restarting to confirm solves the issue as it re-encrypts it with the new key.

**Reply by jensk22** [reply]
Ventz, 'Adding the new API access+refresh tokens' what does that mean?

I have the problem, that I upgraded to 1.27, saw the message about adding ENCRYPTION_KEY to env and first added it to yaml. This eliminated this from the log file, which seemed proper:

2022-09-11 16:44:11.501 [info] Restored encryption key from /opt/app/import/tm_encryption.key:
[36mteslamate_1  |[0m 
[36mteslamate_1  |[0m xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
[36mteslamate_1  |[0m 
[36mteslamate_1  |[0m 2022-09-11 16:44:12.091 [info] POST https://auth.tesla.com/oauth2/v3/token -> 200 (540.266 ms)

However, bot starting with and without the env variable set in ENV, caus ethe logging to stop after a few errors:

[36mteslamate_1  |[0m 2022-09-11 17:14:16.962 [warning] TeslaApi.Error / %{"error" => "{\"error\": \"timeout\"}", "error_description" => "", "response" => nil}
[36mteslamate_1  |[0m 2022-09-11 17:14:16.962 car_id=1 [error] Error / :unknown
[36mteslamate_1  |[0m 2022-09-11 17:14:44.739 car_id=1 [info] Suspending logging

What do I do to  'Adding the new API access+refresh tokens'
Is it entering new tokens using the Andoid (or other) app?

**Reply by ventz** [reply]
@jensk22 Referring to the web UI (see @Etigro's screenshot - although his is in French I believe).

It will ask for the new access+refresh tokens for the Tesla API, for example using something like: https://github.com/adriankumpf/tesla_auth

**Comment by myModelY** [comment]
after insert the key i got this message.

yaml.scanner.ScannerError: while scanning for the next token
found character '\t' that cannot start any token
  in "./docker-compose.yml", line 8, column 1

**Reply by ngardiner** [reply]
You can't use tabs in the docker-compose file, you need to change the tab on line 8 to spaces

**Reply by myModelY** [reply]
i dont have tabs. only space
Encryption_Key=12345678123456781234567812345678

**Reply by ventz** [reply]
You have an “invisible” tab somewhere — if you use vi/vim you may have tabs instead of spaces/etc. Delete line 8 and add it again. 

But also - don’t put passwords/keys in the composer file. Add the variable to the .env file with the secret/password, and reference the variable in the docker-compose file. See the way the DB password/secrets are done.

**Reply by Marcoqwerty78** [reply]
> You have an “invisible” tab somewhere — if you use vi/vim you may have tabs instead of spaces/etc. Delete line 8 and add it again.
> 
> But also - don’t put passwords/keys in the composer file. Add the variable to the .env file with the secret/password, and reference the variable in the docker-compose file. See the way the DB password/secrets are done.

Sorry could you tell me where these .env files are stored? probably i have the same issue...i edited the composer file but with text editor and it split the key on two row. I think on this way doest work
<img width="1361" alt="Schermata 2022-09-20 alle 21 03 57" src="https://user-images.githubusercontent.com/39671834/191344502-4e2b07c0-d2cf-4d83-b7f2-79f898c35e6d.png">

**Reply by ventz** [reply]
@Marcoqwerty78 It's in the same dir (by default) as your dockerc-compose.yml

Start here: https://docs.docker.com/compose/environment-variables/

**Reply by Marcoqwerty78** [reply]
> @Marcoqwerty78 It's in the same dir (by default) as your dockerc-compose.yml
> 
> Start here: https://docs.docker.com/compose/environment-variables/

I have the compose file on the desktop, and there is any other file there. sorry but im totally noob about Docker stuff

**Comment by jensk22** [comment]
I appreciate your answer - but why is this the fix for my problem?

I ONLY updated from 1.26 to 1.27!

regards
Jens

________________________________

**Comment by jensk22** [comment]
It is not that I didn't dare to run the untrusted .exe:

C:\Users\jensk\Downloads>tesla_auth.exe
Error: WebView2 error: WindowsError(Error { code: 0x80070002, message: The system cannot find the file specified.
, win32_error: 2 })

I have read the linked github post/put about tesla-auth!

I don't get why I need this for upgrading from TeslaMate 1.26 to 1.27, but being politite I tried it out. Site informs that one need an Insider Edge build:
Windows

WebView2 is powered by Microsoft Edge (Chromium). At the moment it requires a preview version of Edge which can be downloaded here: Microsoft Edge Insider Channels<https://www.microsoftedgeinsider.com/en-us/download>

And then links to this:

[cid:8bd808ec-1a02-4888-a769-7a088aea1305]

I propose that this  is severely insider nerdish and looks like security by obscufation?

Sorry for being negative, but I just spent 4 hours NOT learning good about Docker/Ubuntu. . . .

regards
Jensk22

________________________________

**Comment by okamiokami** [comment]
This will probably sound stupid, but would it be possible to add a SETTING for this?

I have a Docker installation on a Synology NAS. Even getting to the YML file is a pain in the ass. I copied it, edited, copied back, changed permissions, restarted my NAS, but I keep getting the same error when restarting.

`Conflict. The container name "/docker_grafana_1" is already in use by container "...". You have to remove (or rename) that container to be able to reuse that name.`

What gives?

BTW, I restarted at least 4 times after the update. No login is actually required.

EDIT: updated to Postgres 14, now getting "internal server error". Reverting to Postgres 13 takes me back to the missing encryption key notification. But at least everything else seems to work.

**Comment by hechesspee** [comment]
Hi guys, I'm a non coder so please forgive my ignorance! I'm running teslamate on vultr. getting the below error when I try to login to my teslamate page. Can someone explain me what needs to be done....thanks!
![Error](https://user-images.githubusercontent.com/114410153/192281093-4f84302a-381c-419e-beb9-4c35c0e80071.JPG)

**Reply by cwanja** [reply]
Short and sweet version of it is you need to create (or pull from the logs) the encryption key and put it in your `docker-compose.yml` file. Then restart the containers.

**Reply by Marcoqwerty78** [reply]
> Short and sweet version of it is you need to create (or pull from the logs) the encryption key and put it in your `docker-compose.yml` file. Then restart the containers.

Yes, i think most of the people here try to do it, create key its the easy step, but still there are trouble to paste it somewhere. Also after follow docs.

**Reply by hechesspee** [reply]
Is there a specific location I need to put the encryption key in the docker-compose.yml or could be anywhere?

**Reply by cwanja** [reply]
> Is there a specific location I need to put the encryption key in the docker-compose.yml or could be anywhere?

Add it to the `ENCRYPTION_KEY` variable under services for teslamate. Check out the default yml file on [docs.teslamate.org](https://docs.teslamate.org/docs/installation/docker).

![Screen Shot 2022-09-26 at 8 34 41 AM](https://user-images.githubusercontent.com/54645561/192290429-591a6857-ef27-40b5-a708-ff7928ad0151.png)

**Reply by cwanja** [reply]
@Marcoqwerty78 going through your post here and further up, I am not sure exactly your specific issue. Looks like you were able to paste the encryption key in the right variable for the yml file. What errors are you getting? 

If your TeslaMate instance is not exposed to the internet, you can be less concerned about putting the key in the yml file. If someone has access to the host device (let's say a Raspberry Pi on your home network), you have larger concerns than an encryption key.

**Reply by hechesspee** [reply]
If I'm pulling from the logs is this the command i need to use docker-compose logs teslamate?

**Reply by cwanja** [reply]
Yes.

**Reply by hechesspee** [reply]
I assume I need to restart docker-compose, what is the command? Thanks.

**Reply by cwanja** [reply]
```
docker-compose down
docker-compose up -d
```

**Reply by hechesspee** [reply]
Thanks mate, it worked for me.

**Reply by cwanja** [reply]
Awesome @hechesspee, glad to hear it.

**Reply by Marcoqwerty78** [reply]
> @Marcoqwerty78 going through your post here and further up, I am not sure exactly your specific issue. Looks like you were able to paste the encryption key in the right variable for the yml file. What errors are you getting?
> 
> If your TeslaMate instance is not exposed to the internet, you can be less concerned about putting the key in the yml file. If someone has access to the host device (let's say a Raspberry Pi on your home network), you have larger concerns than an encryption key.

Yes i follow exactly what the doc said, but as you can see on the picture, before posted by me, i still continue to get the error; idk if the Encryption Key must be soo long...but on my case it take two row and probably this cause the error.

I generated the EK by the dedicated ios app.

**Reply by cwanja** [reply]
> Yes i follow exactly what the doc said, but as you can see on the picture, before posted by me, i still continue to get the error; idk if the Encryption Key must be soo long...but on my case it take two row and probably this cause the error.

Is it really on two rows or is Notepads word wrap setting on? Have you tried taking the encryption key out of the logs as mentioned in this reply thread?

**Reply by Marcoqwerty78** [reply]
> > Yes i follow exactly what the doc said, but as you can see on the picture, before posted by me, i still continue to get the error; idk if the Encryption Key must be soo long...but on my case it take two row and probably this cause the error.
> 
> Is it really on two rows or is Notepads word wrap setting on? Have you tried taking the encryption key out of the logs as mentioned in this reply thread?

Yes its true two row, and i didn't try yet to pull out from the log. I will try and i'll let you know. On the doc its written than app its also good.

**Reply by cwanja** [reply]
When you paste it in a new Notepad window, it is two rows? Can you find the line break and hit delete or backspace to remove it? I am sure the app works fine as well.

**Reply by jensk22** [reply]
Just for completeness, I have posted here before and have - as you have - decided to put the encryption key in the yml file, DESPITE being recommended to use an '.env' file instead, which feels right. 

I have read this in completeness: [https://docs.docker.com/compose/environment-variables/]
But can only get ENCRYPTION_KEY to work using the yml file (or by not specifying, it which cause the same log, which is:

[36mteslamate_1  |[0m 2022-09-21 09:54:05.270 [info] Migrations already up
[36mteslamate_1  |[0m 2022-09-21 09:54:08.105 [info] System Info: Erlang/OTP 24 (jit)
[36mteslamate_1  |[0m 2022-09-21 09:54:08.105 [info] Version: 1.27.1
[36mteslamate_1  |[0m 2022-09-21 09:54:08.109 [info] Restored encryption key from /opt/app/import/tm_encryption.key:
[36mteslamate_1  |[0m 
[36mteslamate_1  |[0m PbqnoAo.............................

). 

As I don't get any errors when I write garbage in the .env file, I assume it is simply not used. Don't know if the '.env' file works differently on Windows Docker, would be nice with an '.Env' file example :-)

**Reply by Marcoqwerty78** [reply]
> Short and sweet version of it is you need to create (or pull from the logs) the encryption key and put it in your `docker-compose.yml` file. Then restart the containers.

Ok pulling from the logs works but i saw the key i had generated from the AUTH FOR TESLA ios app its terrible more long than the one temporary pulled out from the log, so i think this app its not working on this case. When i generate the key and paste on the .txt file it will take 2 row
